cuML/src/decisiontree/decisiontree.cu

```diff
@@ -0,0 +1,313 @@
+/*
+ * Copyright (c) 2019, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <utils.h>
+#include "decisiontree.h"
+#include "kernels/gini.cuh"
+#include "kernels/split_labels.cuh"
+#include "kernels/col_condenser.cuh"
+#include "kernels/evaluate.cuh"
+#include "kernels/quantile.cuh"
+
+namespace ML {
+namespace DecisionTree {
+
+template<class T>
+void Question<T>::update(const GiniQuestion<T> & ques) {
+	column = ques.original_column;
+	value = ques.value;
+}
+
+template<class T>
+void TreeNode<T>::print(std::ostream& os) const {
+
+	if (left == nullptr && right == nullptr)
+		os << "(leaf, " << class_predict << ", " << gini_val << ")" ;
+	else
+		os << "(" << question.column << ", " << question.value << ", " << gini_val << ")" ;
+
+	return;
+}
+
+template<typename T>
+std::ostream& operator<<(std::ostream& os, const TreeNode<T> * const node) {
+	node->print(os);
+	return os;
+}
+
+DecisionTreeParams::DecisionTreeParams() {};
```

Comment:
While these methods (and the stateless functions) appear to be part of a public-facing API, we should include doxygen-style comments for completeness. We aren't officially delivering decision-trees, but we do want to make sure we're documenting any APIs we are providing.

Reply:
Yes. The comments are missing in decision tree. We will insert and push them. Similar to what we have in random forest

Reply:
We've added doxygen-style comments both in random-forest and decision trees. Let us know if any additional comments are needed.